### ISSUE TITLE
DexieCloud -> Logout

### ISSUE THREAD

**Issue by b-straub** [dexie/Dexie.js#1689]
Currently the logout of an user is done by calling  "await db.delete();" https://github.com/dexie/Dexie.js/blob/e865e80a3c05821db40ecc3c64157104d345b11e/samples/dexie-cloud-todo-app/src/db/logout.ts#L4

This way causes an _Unhandled rejection: DatabaseClosedError: DatabaseClosedError Database has been closed_
The cause might be, that the LiveQuery from getInternalAccessControlObservable is still active.
https://github.com/dexie/Dexie.js/blob/e865e80a3c05821db40ecc3c64157104d345b11e/addons/dexie-cloud/src/getInternalAccessControlObservable.ts#L22

**Comment by dfahlander** [comment]
Hi! Yes, this is an issue that needs to be fixed in two ways:
1. Provide an easier way to do db.cloud.logout()
2. Generically fix Unhandled rejection bloat. Related issue #1684

**Comment by dfahlander** [comment]
(These were notes to myself about what needs to be done :))

**Comment by b-straub** [comment]
Something is fairly wrong now after the latest changes. When deleting and reopening a database for logout the cache is apparently not purged, even when all observables and live queries are disposed (I can ensure this in my .NET sample app). I think deleting a database should cancel all ongoing subscriptions and purge the cache.

With some slight, probably naive changes (lack of TS knowledge) I could reproduce the problem with the latest toDo sample.

logout.ts
```
import { appComponent } from "../App";
import { CreateDB, DeleteDB } from "./db";

export async function logout() {
  await DeleteDB();
  CreateDB();
  appComponent?.forceUpdate();
}
```

db.ts
```
import { TodoDB } from './TodoDB';

export var db : TodoDB;
let created = false;

export function CreateDB()
{
    if (!created)
    {
        db = new TodoDB();
        created = true;
    }
    
}

export async function DeleteDB()
{
    if (created)
    {
        await db.delete();
        created = false;
    }
}
```

App.tsx
````
import React from 'react';
import { TodoLists } from './components/TodoLists';
import { AddTodoList } from './components/AddTodoList';
import { ResetDatabaseButton } from './components/ResetDatabaseButton';
import { NavBar } from './components/navbar/NavBar';

import 'bootstrap/dist/css/bootstrap.min.css';
import './App.css';
import { Invites } from './components/access-control/Invites';
import { CreateDB } from './db';

interface customWindow extends Window {
  appComponent?: any;
}

export var appComponent : App | undefined; 

class App extends React.Component {

  constructor(props: any) {
    super(props);
    appComponent = this;
    CreateDB();
  }

  forceUpdate() {
    this.setState({});
  }

  render() {
    return (
      <div>
        <NavBar />
        <Invites />
        <TodoLists />
        <AddTodoList />
        <ResetDatabaseButton />
      </div>
    );
  }
}

export default App;
````

**Comment by dfahlander** [comment]
You're right, I've seen the issue too. The optimistic updated cache needs to be purged on database deletion. 
Added as new issue #1736.

**Comment by Baiira** [comment]
Hey!

Just run into this issue, I managed to "fix" it, but I'm not sure if this wouldn't cause a memleak due to some internal dangling pointers, so use at your own discretion (it shouldn't be a problem for us since it's an edge case of an edge case).

It's from an Angular codebase but behaviour should be the reproducible for a React app. Here's some samples:
```
@Injectable({
  providedIn: 'root'
})
export class CollectionsInitializerService {

  constructor(
    private readonly _authCollection: AuthCollection,
    // some others collections...
    private readonly _db: Database,
  ) {
  }

  public reset(): Observable<void> {
    return combineLatest([
      this._authCollection.reset(),
      // call reset on other collections...
    ]).pipe(
      switchMap(() => this._db.remove()),
      map(function () {
        return;
      })
    );
  }
...
}
```


```
type AuthCol = RxCollection<Auth, {}, {}, {}>;

@Injectable({
  providedIn: 'root'
})
export class AuthCollection implements Initializer {
  private _collection?: AuthCol;

  constructor(private readonly db: Database) {
  }

  public reset(): Observable<void> {
    // init is just a wrapper to either use current collection or acquire one from database if _collection is undefined
    return this.init()
      .pipe(
        // not sure about this one, can't hurt I guess?
        switchMap(col => col.remove()),
        map(() => {
          // this is an important one, it seems like collection is not cleared by db.remove()
          this._collection = undefined;
          return;
        })
      );
    // _collection is initialized at some other point in the app lifecycle
  }
...
}
```


```
@Injectable({
  providedIn: 'root'
})
export class Database {
  // used to access database ref from any collections and keep ref
  private _db$ = new BehaviorSubject<RxDatabase | null>(null);

  public remove(): Observable<void> {
    return this._db$.pipe(
      take(1),
      switchMap(db => {
        if (!db) return of(null);
        return db.remove();
      }),
      // recover a new database connection after deletion
      switchMap(() => this.connect()),
      map(() => {
        return;
      })
    );
  }

  private connect(): Observable<unknown> {
    return from(
      createRxDatabase({
        name: DATABASE_NAME,
        storage: getRxStorageDexie(),
      })
    ).pipe(
      take(1),
      tap(db => {
        this._db$.next(db);
      }),
    );
  }
...
}
```

It's a bit more involved than just calling `db.remove()`, but from my testing it seems that collections keep some references of previous entries from removed database and not throwing errors that the db was removed, even if I add some `this._db = undefined`.

I'm using rxdb as a wrapper but this doesn't seems to cause the issue.

Basically, it's just clearing each collections and creating a new one after the database deletion and re-creation of each instances.

I don't have time to try and do pr, maybe in the next months or so, but hopefully it will be changed by then !

Hope this helps, I pruned some lines because their where either not relevant, or tied to business logic.

**Comment by dfahlander** [comment]
There will come a new release dexie@4.0.1-alpha.23 tomorrow where #1736 is fixed and also the general behavior of liveQueries against a database that is deleted and reopened (in same or foreign tab).

**Comment by dfahlander** [comment]
Use latest versions of dexie and dexie-cloud-addon to see these things fixed:

* dexie@4.0.1-alpha.23
* dexie-cloud-addon@4.0.1-beta.45

The todo-app sample is upgraded and republished to https://dexie.github.io/Dexie.js/dexie-cloud-todo-app/
